### PR TITLE
[clang][analyzer] Add function 'fprintf' to StreamChecker.

### PR DESCRIPTION
**Pull request by balazske**
[clang][analyzer] Add function 'fprintf' to StreamChecker.

### CODE REVIEW

**Review by NagyDonat — APPROVED** [review]
I don't see any issue.

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 1]
```diff
@@ -926,6 +932,49 @@ void StreamChecker::evalFputx(const FnDescription *Desc, const CallEvent &Call,
...
+  State = State->BindExpr(CE, C.getLocationContext(), RetVal);
+  SValBuilder &SVB = C.getSValBuilder();
+  auto &ACtx = C.getASTContext();
+  auto Cond = SVB.evalBinOp(State, BO_GE, RetVal, SVB.makeZeroVal(ACtx.IntTy),
+                            SVB.getConditionType())
+                  .getAs<DefinedOrUnknownSVal>();
+  if (!Cond)
+    return;
+  ProgramStateRef StateNotFailed, StateFailed;
+  std::tie(StateNotFailed, StateFailed) = State->assume(*Cond);
```
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:965]: Downstream we used `State->assumeInclusiveRange()` for this.

As a sideffect, we bound an upperbound to the `RetVal` with the number of arguments we had for the call.
Have you considered setting the upperbound here too?

I was thinking if the std library functions checker applies this constraint, but AFAIK it does not, nor we have a proposed patch for doing so. And anyways I'd prefer setting both upper and lowerbounds at the same checker if possible. Thus, I write to this PR.

Comment (balazs-benics-sonarsource) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:965]: In addition to this, we should also invalidate any buffers passed to such calls. This also sounds like a regression compared to default eval calling "fpintf". (Look at the format string specifier `%n`)

Comment (balazske) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:965]: Yes, some fixes can be done with `fprintf` and `fscanf` and the argument invalidation. I plan to do this with the following patches.

**Pull request merged**